### PR TITLE
different conversion ENU to LLA, added noise to position

### PR DESCRIPTION
**Pull request by ptrkfry**
Computing the WGS-84 coordinates from the Gazebo position yields an error in one axis if using the conversion implemented in Gazebo (at least for Gazebo 5, I haven't tested other versions). 
That is why I added "geodetic_conv.hpp" from the package geodetic_utils to do the conversion instead. It would probably make more sense to get this directly from the package geodetic_utils and add this package as a dependency, but I wasn't sure how to do this.
Also I added noise on the GPS position measurement. Before, we could define a variance for the position, but no noise was added (only to the ground speed). The case variance = 0 should be checked, the noise seems to be not exactly zero..not sure why

### CODE REVIEW

**Comment by hitimo** [comment]
add to whitelist

**Comment by hitimo** [comment]
@pvechersky That's what you noticed a while back, correct?

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 1]
```diff
@@ -190,9 +194,40 @@ void GazeboGpsPlugin::OnUpdate() {
...
+  // Get the world pose of the sensor.
+  std::string parentName = parent_sensor_->GetParentName();
+  physics::EntityPtr parent = world_->GetEntity(parentName);
+  math::Pose sensorPose = parent_sensor_->GetPose() + parent->GetWorldPose();
+
+  // Get the position out of the pose.
+  math::Vector3 pos = sensorPose.pos;
+
+  // Add noise to position.
+  pos += math::Vector3(position_n_[0](random_generator_),
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: @ptrkfry Add noise only if desired sigma != 0.0 as discussed offline

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 2]
```diff
@@ -175,9 +175,16 @@ void GazeboGpsPlugin::OnUpdate() {
...
   // Apply noise to ground speed.
-  W_ground_speed_W_L += math::Vector3(ground_speed_n_[0](random_generator_),
-                                      ground_speed_n_[1](random_generator_),
-                                      ground_speed_n_[2](random_generator_));
+  if(hor_vel_std_dev_ != 0) {
+    W_ground_speed_W_L.Set(W_ground_speed_W_L[0] + ground_speed_n_[0](random_generator_),
+                           W_ground_speed_W_L[1] + ground_speed_n_[1](random_generator_),
+                           W_ground_speed_W_L[2]);
+  }
+  if(ver_vel_std_dev_ != 0) {
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: `if (ver_vel_std_dev_ > 0.0) {`

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 3]
```diff
@@ -190,9 +197,47 @@ void GazeboGpsPlugin::OnUpdate() {
...
-  gz_gps_message_.set_latitude(parent_sensor_->GetLatitude().Degree());
-  gz_gps_message_.set_longitude(parent_sensor_->GetLongitude().Degree());
-  gz_gps_message_.set_altitude(parent_sensor_->GetAltitude());
+  // Get the world pose of the sensor.
+  std::string parentName = parent_sensor_->GetParentName();
+  physics::EntityPtr parent = world_->GetEntity(parentName);
+  math::Pose sensorPose = parent_sensor_->GetPose() + parent->GetWorldPose();
+
+  // Get the position out of the pose.
+  math::Vector3 pos = sensorPose.pos;
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: Really needed? --> Directly use `sensorPose.pos` below or const reference

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 4]
```diff
@@ -190,9 +197,47 @@ void GazeboGpsPlugin::OnUpdate() {
 #else
   current_time = parent_sensor_->GetLastMeasurementTime();
 
-  gz_gps_message_.set_latitude(parent_sensor_->GetLatitude().Degree());
-  gz_gps_message_.set_longitude(parent_sensor_->GetLongitude().Degree());
-  gz_gps_message_.set_altitude(parent_sensor_->GetAltitude());
+  // Get the world pose of the sensor.
+  std::string parentName = parent_sensor_->GetParentName();
+  physics::EntityPtr parent = world_->GetEntity(parentName);
+  math::Pose sensorPose = parent_sensor_->GetPose() + parent->GetWorldPose();
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: Is this correct?

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 5]
```diff
@@ -190,9 +197,47 @@ void GazeboGpsPlugin::OnUpdate() {
...
+
+  // Get the position out of the pose.
+  math::Vector3 pos = sensorPose.pos;
+
+  // Add noise to position.
+  if(hor_pos_std_dev_ != 0) {
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: same

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 6]
```diff
@@ -190,9 +197,47 @@ void GazeboGpsPlugin::OnUpdate() {
...
+    pos.Set(pos[0] + position_n_[0](random_generator_),
+            pos[1] + position_n_[1](random_generator_),
+            pos[2]);
+  }
+  if(ver_pos_std_dev_ != 0) {
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: same

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 7]
```diff
@@ -190,9 +197,47 @@ void GazeboGpsPlugin::OnUpdate() {
...
+
+  // Set up the GPS conversion ENU-->LLA.
+  // Initialise the reference point for the conversion.
+  // Reference point = origin of Gazebo world.
+  geodetic_converter::GeodeticConverter* g = new geodetic_converter::GeodeticConverter;
+  math::Angle lat_ref = world_->GetSphericalCoordinates()->GetLatitudeReference();
+  math::Angle lon_ref = world_->GetSphericalCoordinates()->GetLongitudeReference();
+  double alt_ref = world_->GetSphericalCoordinates()->GetElevationReference();
+  double lat_ref_deg = lat_ref.Degree();
+  double lon_ref_deg = lon_ref.Degree();
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: Consider passing this directly to `initialiseReference`

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 8]
```diff
@@ -190,9 +197,47 @@ void GazeboGpsPlugin::OnUpdate() {
...
+
+  g->initialiseReference(lat_ref_deg, lon_ref_deg, alt_ref);
+  
+  //Convert ENU to LLA (WGS-84).
+  double latitude = 0;
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: `0.0`

Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: latitude_deg

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 9]
```diff
@@ -190,9 +197,47 @@ void GazeboGpsPlugin::OnUpdate() {
...
+  double longitude = 0;
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: same

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 10]
```diff
@@ -190,9 +197,47 @@ void GazeboGpsPlugin::OnUpdate() {
...
+  double altitude = 0;
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: `0.0
`

Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: altitude_m

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp (hunk 4 above)
Comment (ptrkfry) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: It think so.. when I convert the produced GPS coordinates back to the local coordinate system, it yields the same as the ground truth messages.
But yeah there might be a prettier way to do this.. but not sure how.

**Comment by pvechersky** [comment]
@hitimo Sorry for a late response. But yes, this is exactly what I had noticed at one point as well. For my workaround I did some manual calculations to convert this into the geodetic coordinates, similar to how the Hector gazebo gps plugin does it. However, leveraging a designated package for the conversion is probably more robust :)

**Comment by Jaeyoung-Lim** [comment]
Closing as stale, please reopen if this is still of interest

**Pull request closed without merging**